Readme.md
Arduino HID Project 2.8.0
=========================

![Header Picture](header.jpg)

This project went through a lot of phases and has now reached a new Arduino ~~USB-Core~~ Library
with a lot of new functions like extended HID. It also supports HoodLoader2 (version 1 is no longer supported).
The idea is to enable enhanced USB functions to almost all 'standard' Arduino boards.

<a href="https://www.buymeacoffee.com/nicohood" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a>

**Supported Arduinos (IDE 1.6.7 or higher!):**
* Uno (with [HoodLoader2](https://github.com/NicoHood/HoodLoader2))
* Mega (with [HoodLoader2](https://github.com/NicoHood/HoodLoader2))
* Leonardo
* (Pro)Micro
* Any other 8u2/16u2/at90usb8/162/32u2/32u4 compatible board
* Zero
* MKR1000
* Any other Samd21 compatible board
* No ATSAM3 support (Due, etc)

**Supported HID devices:**
* Keyboard with Leds out (8 modifiers + 6 keys pressed at the same time, + 1 limited linux consumer key)
* ~~Teensy Keyboard with different keyboard layouts (german, french and many more)~~ soon
* NKRO Keyboard with Leds out (press up to 113 keys at the same time)
* Mouse (5 buttons, move, wheel)
* BootKeyboard/BootMouse BIOS protocol support
* Absolute Mouse
* Consumer/Media Keys (4 keys for music player, web browser and more)
* System Key (for PC standby/shutdown)
* Gamepad (32 buttons, 4 16bit axis, 2 8bit axis, 2 D-Pads)
* RawHID
* SurfaceDial
* Touchscreen (max 10 touches at the same time)
* Each device is available as single or multi report device (except RawHID and Touchscreen)

See the [wiki](https://github.com/NicoHood/HID/wiki/Features) for more information about features etc.

Wiki
====

All documentation moved to the [wiki page](https://github.com/NicoHood/HID/wiki).

An offline snapshot is available in [releases](https://github.com/NicoHood/HID/releases).


Contact
=======

www.nicohood.de

**If you have any questions, please ask at the [Github Discussions Page](https://github.com/NicoHood/HID/discussions).**


Version History
===============
```
2.8.0 Release (08.04.2021)
* Added lots of keyboard layout languages (Thanks @SukkoPera)

2.7.0 Release (06.03.2021)
* Fix SAMD compilation #244
* Added support for Italien Keymap #231
* Changed: Use a fixed definion of keycodes and add language specific variants

2.6.2 Release (24.12.2020)
* Fix SAMD alignment #211
* Fix asciimap size check #228

2.6.1 Release (13.07.2019)
* Allow keyboard layout change within arduino sketch #176

2.6.0 Release (16.04.2019)
* Added generic keyboard layout support #168

2.5.0 Release (15.09.2018)
* Ported to SAMD architecture #146
* Fixed bad keyboard report issue #148

2.4.4 Release (27.01.2017)
* Added releaseAll() to Mouse API
* Fix flexible array errors

2.4.3 Release (02.03.2016)
* Fixed NKRO Keyboard modifier add() #76

2.4.2 Release (04.01.2016)
* RawHID >64 byte fix (Arduino IDE 1.6.8 fix)
* Fixed Keyboard modifier add() issue #68

2.4.1 Release (19.12.2015)
* RawHID Improvements
* Added Keyboard Feature Report
* NKRO and Keyboard API fixes
* Arduino library manager fix

2.4 Release (06.11.2015)
* Added Arduino IDE 1.6.6 compatibility with Pluggable HID
* Improved Pluggable HID (see Arduyuino changelog for my improvements)
* Changed USB-Core into a simple library, only made possible with Pluggable HID
* Removed HID presets in boards menu (like mouse + keyboard + consumer + system)
* Added NKRO Keyboard
* Added Led report for Keyboard
* Added 1 Linux consumer key for keyboard
* Added BootKeyboard/Mouse support (BIOS compatibility)
* Added RawHID
* Added a few key definitions
* Renew whole Keyboard API and its definitions via enum
* Uses .alinkage custom IDE option
* Improved and updated examples
* A lot of other minor and major fixes I missed to mention.

2.3 Release (never released)
* Updated Libraries
* Updated Arduino Core
* Added Minor Consumer definitions
* Fixed platforms.txt
* SERIAL_RX_BUFFER_SIZE reverted to 16 (TODO add -D to build option)

2.2 Release (12.04.2015)
* added experimental, not finished nor documented HID-Bridge between 16u2 and 328/2560
* increased HW Serial1 RX buffer size from 16 to 32 (TX still 16)
* added colour highlighting (through HID-Bridge library)
* removed fixed size in report buffers
* used HID_KeyboardReport_Data_t now in Keyboard API
* Arduino as ISP fix for 328
* Upload verification on USB hubs fix for HL2.0.4
* No USB workaround for Leonardo integrated into variants
* Changed USB Wakeup in System API
* Consumer Key fix (issue #3)
* Gamepad fix (issue #14)
* Added Keycode example
* Mouse press + release fix

2.1 Release (28.01.2015)
* Reworked the whole USB-Core from scratch
* Uses less flash if HID or Serial is not used
* Extended and compacter(flash) HID Report Descriptors
* Fixed USB Device Descriptor
* Added u2 compatibility with smaller USB_EP_SIZE (16u2 etc)
* Added Serial Event for LineEncoding and LineState
* Added Serial Function to get dtr state, line encoding etc
* Added Keyboard Led Out report to get Led states (for numlock etc)
* Made CDC-Core independent from USB-Core
* Made HID-Core independent from USB-Core
* Made HID-API independent from HID-Core
* Removed not needed virtual functions in Keyboard
* Made HID Reports and its IDs replaceable via pins_Arduino.h
* Added Absolute Mouse
* Removed uint8_t USBPutChar(uint8_t c); in HID.cpp
* Made void Recv(volatile u8* data, u8 count) in USBCore.cpp static inline
* HID-APIs sends a clean report on begin() and end() now.
* Removed virtual functions in Keyboard API
* Added Keycode functions in Keyboard API
* Inlined a lot of the HID API functions to save flash
* Added Gamepad
* Added RawHID API (but RawHID itself isnt working still)
* Added USB Wakeup support
* Separated USB-Core in its own folder
* Added HID Tables
* USB-Serial now fully reprogrammable
* Easy USB-Core selection via Tools->USB-Core
* Added Arduino as ISP fix for 32u4 and 16u2
* Moved documentation to the wiki
* Added AVR libraries to the core with the better SoftSerial

2.0 Release (29.11.2014)
* Added HoodLoader2
* Separated HoodLoader1&2 more
* Added u2 series for USB-Core
* Extended USB core and fixed minor things for the u2 series
* Added Led Out report.
* Added CDC Line state
* Reworked the whole library structure again

1.8 Beta Release (26.08.2014)
* Changes in the Hoodloader1:
 * **Huge improvements**, see Hoodloader1 repository
 * Reworked the whole library, easy installation now
 * HID fixes for Media Keys/Ubuntu
 * Removed Joystick, added 4 Gamepads

1.7.3 Beta Release (10.08.2014)
* Changes in the Hoodloader1:
 * Fixed HID flush bug (1.6 - 1.7.2)

1.7.2 Beta Release (10.08.2014)
* Changes in the Hoodloader1:
 * Added Lite version for 8u2
 * Added Versions that show up as Uno/Mega (not recommended)
 * Makefile and structure changes

1.7.1 Beta Release (10.08.2014)
* Changes in the Hoodloader1:
 * Fixed HID deactivation bug

1.7 Beta Release (10.08.2014)
* Changes in the Hoodloader1:
 * Works as ISP now. See Hoodloader1 Repository for more information.
 * Exceeded 8kb limit. For flashing a 8u2 use v1.6 please!
* Changed Readme text

1.6 Beta Release (09.08.2014)
* Bugfixes in the Hoodloader1:
 * Changed HID management (not blocking that much, faster)
 * added RawHID in/out (HID to Serial)
* Added RawHID Class and example

1.5 Beta Release (21.07.2014)
* Moved Hoodloader1 source to a separate Github page
* Bugfixes in the Hoodloader:
 * Firmware is still available here
 * Overall a lot of ram improvements, now with a big global union of ram
 * Removed USBtoUSART buffer (not needed, saved 128/500 bytes)
 * Removed Lite version because of better ram usage not needed
 * Separated different modes better to not cause any errors in default mode
 * Improved the deactivate option
 * Integrated NHP directly
 * Replaced LightweightRingbuffer with native Lufa Ringbuffer
 * Improved writing to CDC Host
 * Fixed a bug in checkNHPProtocol: & needs to be a ==
 * General structure changes
 * Improved stability
 * Fixed Arduino as ISP bug

1.4.1 Beta Release (10.07.2014)
* #define Bugfix in USBAPI.h

1.4 Beta Release (10.07.2014)
* Bugfixes in the Hoodloader1:
 * Added Lite Version with less ram usage
 * Changed PIDs, edited driver file
* merged v1.0.x and v1.5.x together (both are compatible!)
* added IDE v1.5.7 support
* added Tutorials

1.3 Beta Release (01.07.2014)
* Bugfixes in the Hoodloader1:
 * Improved ram usage
* **Important NHP fix inside the HID Class for Uno/Mega**

1.2 Beta Release (22.06.2014)
* Added 1.0.x/1.5.x support
* Bugfixes in the Hoodloader1:
 * Sometimes HID Devices weren't updating
   when using more than 1 Device (set forcewrite to true)
 * Fast updates crashed the bootloader
  (too much ram usage, set CDC buffer from 128b to 100b each)
* Minor file structure changes

1.1 Beta Release (05.06.2014)
* Added Leonardo/Micro support
* Included NicoHoodProtocol
* Minor fixes

1.0 Beta Release (03.06.2014)
```


Licence and Copyright
=====================
If you use this library for any cool project let me know!

```
Copyright (c) 2014-2016 NicoHood
See the readme for credit to other people.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
This changlog uses the [ISO 8601 date format](https://www.iso.org/iso-8601-date-and-time-format.html) of (YYYY-MM-DD).

## [Unreleased]

## [2.8.4] - 2022-09-23

### Fixed

* Fix `library.properties`, so the latest version will appear in the library manager.

## [2.8.3] - 2022-01-11

### Added

* Japanese Keyboard Layout [#332](https://github.com/NicoHood/HID/pull/332) (Thanks [@HisashiKato](https://github.com/HisashiKato))

### Fixed

* `SET_IDLE` uses wValueH for duration, not wValueL [#310](https://github.com/NicoHood/HID/pull/310) (Thanks [@mdevaev](https://github.com/mdevaev))
* Handle `HID_HID_DESCRIPTOR_TYPE` properly [#311](https://github.com/NicoHood/HID/pull/311) (Thanks [@mdevaev](https://github.com/mdevaev))
* Missing return value added to samd core [#342](https://github.com/NicoHood/HID/pull/342) (Thanks [@ladyada](https://github.com/ladyada))
* Fixed German Layout [#323](https://github.com/NicoHood/HID/pull/323) (Thanks [@jensweimann](https://github.com/jensweimann))

## [2.8.2] - 2021-07-14

### Fixed

* Improved Windows 7 Absolute Mouse compatibility [#306](https://github.com/NicoHood/HID/issues/305) (Thanks [@mdevaev](https://github.com/mdevaev))

## [2.8.1] - 2021-07-14

### Changed

* Moved License from readme to its own file
* Moves this changelog from readme into its own file (following keepachangelog format)

### Fixed
* Improved Mac Mouse compatibility [#305](https://github.com/NicoHood/HID/issues/305) (Thanks [@mdevaev](https://github.com/mdevaev))

## [2.8.0] - 2021-04-08

### Added

* Added lots of keyboard layout languages (Thanks [@SukkoPera](https://github.com/SukkoPera))

## [2.7.0] - 2021-03-06

### Added

* Added support for Italien Keymap [#231](https://github.com/NicoHood/HID/issues/231)

### Changed

* Changed: Use a fixed definion of keycodes and add language specific variants

### Fixed

* Fix SAMD compilation [#244](https://github.com/NicoHood/HID/issues/244)

## [2.6.2] - 2020-12-24

### Fixed

* Fix SAMD alignment [#211](https://github.com/NicoHood/HID/issues/211)
* Fix asciimap size check [#228](https://github.com/NicoHood/HID/issues/228)

## [2.6.1] - 2019-07-13

### Added

* Allow keyboard layout change within arduino sketch [#176](https://github.com/NicoHood/HID/issues/176)

## [2.6.0] - 2019-04-16

### Added

* Added generic keyboard layout support [#168](https://github.com/NicoHood/HID/issues/168)

## [2.5.0] - 2018-09-15

### Added

* Ported to SAMD architecture [#146](https://github.com/NicoHood/HID/issues/146)

### Fixed

* Fixed bad keyboard report issue [#148](https://github.com/NicoHood/HID/issues/148)


## [2.4.4] - 2017-01-27

### Added

* Added `releaseAll()` to Mouse API

### Fixed

* Fix flexible array errors


## [2.4.3] - 2016-03-02

### Fixed

* Fixed NKRO Keyboard modifier add() [#76](https://github.com/NicoHood/HID/issues/76)


## [2.4.2] - 2015-12-19

### Fixed

* RawHID >64 byte fix (Arduino IDE 1.6.8 fix)
* Fixed Keyboard modifier add() issue [#68](https://github.com/NicoHood/HID/issues/68)

## [2.4.1] - 2015-12-19

### Added

* Added Keyboard Feature Report

### Changed

* RawHID Improvements

### Fixed

* NKRO and Keyboard API fixes
* Arduino library manager fix

## [2.4] - 2015-11-06

### Added

* Added Arduino IDE 1.6.6 compatibility with Pluggable HID
* Added NKRO Keyboard
* Added Led report for Keyboard
* Added 1 Linux consumer key for keyboard
* Added BootKeyboard/Mouse support (BIOS compatibility)
* Added RawHID
* Added a few key definitions
* Uses .alinkage custom IDE option

### Changed

* Improved Pluggable HID (see Arduyuino changelog for my improvements)
* Changed USB-Core into a simple library, only made possible with Pluggable HID
* Renew whole Keyboard API and its definitions via enum
* Improved and updated examples

### Removed

* Removed HID presets in boards menu (like mouse + keyboard + consumer + system)

### Fixed

* A lot of other minor and major fixes I missed to mention.

## 2.3 (Never released)

### Added

* Added Minor Consumer definitions

### Changed

* Updated Libraries
* Updated Arduino Core
* `SERIAL_RX_BUFFER_SIZE` reverted to 16 (TODO add -D to build option)

### Fixed

* Fixed platforms.txt

## [2.2] - 2015-04-12

### Added

* Added experimental, not finished nor documented HID-Bridge between 16u2 and 328/2560
* Added colour highlighting (through HID-Bridge library)
* No USB workaround for Leonardo integrated into variants
* Added Keycode example

### Changed

* Increased HW Serial1 RX buffer size from 16 to 32 (TX still 16)
* Used `HID_KeyboardReport_Data_t` now in Keyboard API
* Changed USB Wakeup in System API

### Removed

* Removed fixed size in report buffers

### Fixed

* Arduino as ISP fix for 328
* Upload verification on USB hubs fix for HL2.0.4
* Consumer Key fix (issue [#3](https://github.com/NicoHood/HID/issues/3))

HID.h
/*
HID.h
Copyright (c) 2005-2014 Arduino.  All right reserved.

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef __HID__
#define __HID__

#include "Arduino.h"

// enable Led out report by default
#ifndef EXTERN_HID_REPORT
#define HID_KEYBOARD_LEDS_ENABLED
#endif

// HID report IDs
// note by NicoHood: I would not change the current IDs, since it can confuse the OS
// I had several problems if i change the report id and its use.
// It doesnt matter if a device is not presented and there is a lack of IDs (at least number 1 must be represented)
// Report IDs and the report itself can be overwritten by the pins_arduino.h
#ifndef HID_REPORTID_MOUSE
#define HID_REPORTID_MOUSE 1
#endif
#ifndef HID_REPORTID_KEYBOARD
#define HID_REPORTID_KEYBOARD 2
#endif
#ifndef HID_REPORTID_RAWHID
#define HID_REPORTID_RAWHID 3
#endif
#ifndef HID_REPORTID_CONSUMERCONTROL
#define HID_REPORTID_CONSUMERCONTROL 4
#endif
#ifndef HID_REPORTID_SYSTEMCONTROL
#define HID_REPORTID_SYSTEMCONTROL 5
#endif
#ifndef HID_REPORTID_GAMEPAD
#define HID_REPORTID_GAMEPAD 6
#endif
#ifndef HID_REPORTID_MOUSE_ABSOLUTE
#define HID_REPORTID_MOUSE_ABSOLUTE 7
#endif

// HID reports
// Report IDs and the report itself can be overwritten by the pins_arduino.h


#ifndef HID_REPORT_MOUSE
#define HID_REPORT_MOUSE(report_id) /*  Mouse relative */ \
    0x05, 0x01,                      /* USAGE_PAGE (Generic Desktop)	  54 */ \
    0x09, 0x02,                      /* USAGE (Mouse) */ \
    0xa1, 0x01,                      /* COLLECTION (Application) */ \
    0x85, report_id,				/*     REPORT_ID */ \
\
	/* 8 Buttons */ \
    0x05, 0x09,                      /*     USAGE_PAGE (Button) */ \
    0x19, 0x01,                      /*     USAGE_MINIMUM (Button 1) */ \
    0x29, 0x08,                      /*     USAGE_MAXIMUM (Button 8) */ \
    0x15, 0x00,                      /*     LOGICAL_MINIMUM (0) */ \
    0x25, 0x01,                      /*     LOGICAL_MAXIMUM (1) */ \
    0x95, 0x08,                      /*     REPORT_COUNT (8) */ \
    0x75, 0x01,                      /*     REPORT_SIZE (1) */ \
    0x81, 0x02,                      /*     INPUT (Data,Var,Abs) */ \
\
	/* X, Y, Wheel */ \
    0x05, 0x01,                      /*     USAGE_PAGE (Generic Desktop) */ \
    0x09, 0x30,                      /*     USAGE (X) */ \
    0x09, 0x31,                      /*     USAGE (Y) */ \
    0x09, 0x38,                      /*     USAGE (Wheel) */ \
    0x15, 0x81,                      /*     LOGICAL_MINIMUM (-127) */ \
    0x25, 0x7f,                      /*     LOGICAL_MAXIMUM (127) */ \
    0x75, 0x08,                      /*     REPORT_SIZE (8) */ \
    0x95, 0x03,                      /*     REPORT_COUNT (3) */ \
    0x81, 0x06,                      /*     INPUT (Data,Var,Rel) */ \
\
	/* End */ \
    0xc0                            /* END_COLLECTION */ 
#endif

#ifndef HID_REPORT_MOUSE_ABSOLUTE
#define HID_REPORT_MOUSE_ABSOLUTE(report_id) /*  Mouse absolute */ \
	0x05, 0x01,							/* Usage Page (Generic Desktop) */ \
	0x09, 0x02,							/* Usage (Mouse) */ \
	0xA1, 0x01,							/* Collection (Application) */ \
	0x85, report_id,					/* REPORT_ID */ \
	0x05, 0x01,							/* Usage Page (Generic Desktop) */ \
	0x09, 0x30,							/* Usage (X) */ \
	0x09, 0x31,							/* Usage (Y) */ \
	0x15, 0x00,							/* Logical Minimum (0) */ \
	0x26, 0xFF, 0x7F,					/* Logical Maximum (32767) */ \
	0x75, 0x10,							/* Report Size (16), */ \
	0x95, 0x02,							/* Report Count (2), */ \
	0x81, 0x02,							/* Input (Data, Variable, Absolute) */ \
	0xC0								/* End Collection */
#endif

#ifndef HID_REPORT_KEYBOARD_LEDS
#define HID_REPORT_KEYBOARD_LEDS(report_id)    /*  Keyboard */ \
    0x05, 0x01,                      /* USAGE_PAGE (Generic Desktop)	  47 */ \
    0x09, 0x06,                      /* USAGE (Keyboard) */ \
    0xa1, 0x01,                      /* COLLECTION (Application) */ \
    0x85, report_id,				 /*   REPORT_ID */ \
    0x05, 0x07,                      /*   USAGE_PAGE (Keyboard) */ \
\
      /* Keyboard Modifiers (shift, alt, ...) */ \
    0x19, 0xe0,                      /*   USAGE_MINIMUM (Keyboard LeftControl) */ \
    0x29, 0xe7,                      /*   USAGE_MAXIMUM (Keyboard Right GUI) */ \
    0x15, 0x00,                      /*   LOGICAL_MINIMUM (0) */ \
    0x25, 0x01,                      /*   LOGICAL_MAXIMUM (1) */ \
    0x75, 0x01,                      /*   REPORT_SIZE (1) */ \
	0x95, 0x08,                      /*   REPORT_COUNT (8) */ \
    0x81, 0x02,                      /*   INPUT (Data,Var,Abs) */ \
\
      /* Reserved byte */ \
    0x95, 0x01,                      /*   REPORT_COUNT (1) */ \
    0x75, 0x08,                      /*   REPORT_SIZE (8) */ \
    0x81, 0x03,                      /*   INPUT (Cnst,Var,Abs) */ \
\
	/* 5 LEDs for num lock etc */ \
	0x05, 0x08,						 /*   USAGE_PAGE (LEDs) */ \
	0x19, 0x01,						 /*   USAGE_MINIMUM (Num Lock) */ \
	0x29, 0x05,						 /*   USAGE_MAXIMUM (Kana) */ \
	0x95, 0x05,						 /*   REPORT_COUNT (5) */ \
	0x75, 0x01,						 /*   REPORT_SIZE (1) */ \
	0x91, 0x02,						 /*   OUTPUT (Data,Var,Abs) */ \
	/*  Reserved 3 bits */ \
	0x95, 0x01,						 /*   REPORT_COUNT (1) */ \
	0x75, 0x03,						 /*   REPORT_SIZE (3) */ \
	0x91, 0x03,						 /*   OUTPUT (Cnst,Var,Abs) */ \
\
      /* 6 Keyboard keys */ \
    0x95, 0x06,                      /*   REPORT_COUNT (6) */ \
    0x75, 0x08,                      /*   REPORT_SIZE (8) */ \
    0x15, 0x00,                      /*   LOGICAL_MINIMUM (0) */ \
    0x26, 0xE7, 0x00,                /*   LOGICAL_MAXIMUM (231) */ \
    0x05, 0x07,                      /*   USAGE_PAGE (Keyboard) */ \
    0x19, 0x00,                      /*   USAGE_MINIMUM (Reserved (no event indicated)) */ \
    0x29, 0xE7,                      /*   USAGE_MAXIMUM (Keyboard Right GUI) */ \
    0x81, 0x00,                      /*   INPUT (Data,Ary,Abs) */ \
\
      /* End */ \
    0xc0                            /* END_COLLECTION */
#endif

#ifndef HID_REPORT_KEYBOARD_KEYS
#define HID_REPORT_KEYBOARD_KEYS(report_id)    /*  Keyboard */ \
    0x05, 0x01,                      /* USAGE_PAGE (Generic Desktop)	  47 */ \
    0x09, 0x06,                      /* USAGE (Keyboard) */ \
    0xa1, 0x01,                      /* COLLECTION (Application) */ \
    0x85, report_id,				 /*   REPORT_ID */ \
    0x05, 0x07,                      /*   USAGE_PAGE (Keyboard) */ \
\
      /* Keyboard Modifiers (shift, alt, ...) */ \
    0x19, 0xe0,                      /*   USAGE_MINIMUM (Keyboard LeftControl) */ \
    0x29, 0xe7,                      /*   USAGE_MAXIMUM (Keyboard Right GUI) */ \
    0x15, 0x00,                      /*   LOGICAL_MINIMUM (0) */ \
    0x25, 0x01,                      /*   LOGICAL_MAXIMUM (1) */ \
    0x75, 0x01,                      /*   REPORT_SIZE (1) */ \
	0x95, 0x08,                      /*   REPORT_COUNT (8) */ \
    0x81, 0x02,                      /*   INPUT (Data,Var,Abs) */ \
\
      /* Reserved byte */ \
    0x95, 0x01,                      /*   REPORT_COUNT (1) */ \
    0x75, 0x08,                      /*   REPORT_SIZE (8) */ \
    0x81, 0x03,                      /*   INPUT (Cnst,Var,Abs) */ \
\
	HID_REPORT_KEYBOARD_LEDS \
      /* 6 Keyboard keys */ \
    0x95, 0x06,                      /*   REPORT_COUNT (6) */ \
    0x75, 0x08,                      /*   REPORT_SIZE (8) */ \
    0x15, 0x00,                      /*   LOGICAL_MINIMUM (0) */ \
    0x26, 0xE7, 0x00,                /*   LOGICAL_MAXIMUM (231) */ \
    0x05, 0x07,                      /*   USAGE_PAGE (Keyboard) */ \
    0x19, 0x00,                      /*   USAGE_MINIMUM (Reserved (no event indicated)) */ \
    0x29, 0xE7,                      /*   USAGE_MAXIMUM (Keyboard Right GUI) */ \
    0x81, 0x00,                      /*   INPUT (Data,Ary,Abs) */ \
\
      /* End */ \
    0xc0                            /* END_COLLECTION */
#endif

//TODO limit to system keys only?
#ifndef HID_REPORT_SYSTEMCONTROL
#define HID_REPORT_SYSTEMCONTROL(report_id) /*  System Control (Power Down, Sleep, Wakeup, ...) */ \
	0x05, 0x01,								/* USAGE_PAGE (Generic Desktop) */ \
	0x09, 0x80,								/* USAGE (System Control) */ \
	0xa1, 0x01, 							/* COLLECTION (Application) */ \
	0x85, report_id,						/* REPORT_ID */ \
	/* 1 system key */ \
	0x15, 0x00, 							/* LOGICAL_MINIMUM (0) */ \
	0x26, 0xff, 0x00, 						/* LOGICAL_MAXIMUM (255) */ \
	0x19, 0x00, 							/* USAGE_MINIMUM (Undefined) */ \
	0x29, 0xff, 							/* USAGE_MAXIMUM (System Menu Down) */ \
	0x95, 0x01, 							/* REPORT_COUNT (1) */ \
	0x75, 0x08, 							/* REPORT_SIZE (8) */ \
	0x81, 0x00, 							/* INPUT (Data,Ary,Abs) */ \
	0xc0 									/* END_COLLECTION */
#endif

#ifndef HID_REPORT_CONSUMERCONTROL
#define HID_REPORT_CONSUMERCONTROL(report_id)	/* Consumer Control (Sound/Media keys) */ \
	0x05, 0x0C,									/* usage page (consumer device) */ \
	0x09, 0x01, 								/* usage -- consumer control */ \
	0xA1, 0x01, 								/* collection (application) */ \
	0x85, report_id, 							/* report id */ \
	/* 4 Media Keys */ \
	0x15, 0x00, 								/* logical minimum */ \
	0x26, 0xFF, 0x03, 							/* logical maximum (3ff) */ \
	0x19, 0x00, 								/* usage minimum (0) */ \
	0x2A, 0xFF, 0x03, 							/* usage maximum (3ff) */ \
	0x95, 0x04, 								/* report count (4) */ \
	0x75, 0x10, 								/* report size (16) */ \
	0x81, 0x00, 								/* input */ \
	0xC0 /* end collection */
#endif

#ifndef HID_REPORT_GAMEPAD
#define HID_REPORT_GAMEPAD(report_id) /* Gamepad with 32 buttons and 6 axis*/ \
	0x05, 0x01,							/* USAGE_PAGE (Generic Desktop) */ \
	0x09, 0x04,							/* USAGE (Joystick) */ \
	0xa1, 0x01,							/* COLLECTION (Application) */ \
	0x85, HID_REPORTID_GAMEPAD,	/*   REPORT_ID */ \
	/* 32 Buttons */ \
	0x05, 0x09,							/*   USAGE_PAGE (Button) */ \
	0x19, 0x01,							/*   USAGE_MINIMUM (Button 1) */ \
	0x29, 0x20,							/*   USAGE_MAXIMUM (Button 32) */ \
	0x15, 0x00,							/*   LOGICAL_MINIMUM (0) */ \
	0x25, 0x01,							/*   LOGICAL_MAXIMUM (1) */ \
	0x75, 0x01,							/*   REPORT_SIZE (1) */ \
	0x95, 0x20,							/*   REPORT_COUNT (32) */ \
	0x81, 0x02,							/*   INPUT (Data,Var,Abs) */ \
	/* 4 16bit Axis */ \
	0x05, 0x01,							/*   USAGE_PAGE (Generic Desktop) */ \
	0xa1, 0x00,							/*   COLLECTION (Physical) */ \
	0x09, 0x30,							/*     USAGE (X) */ \
	0x09, 0x31,							/*     USAGE (Y) */ \
	0x09, 0x33,							/*     USAGE (Rx) */ \
	0x09, 0x34,							/*     USAGE (Ry) */ \
	0x16, 0x00, 0x80,					/*     LOGICAL_MINIMUM (-32768) */ \
	0x26, 0xFF, 0x7F,					/*     LOGICAL_MAXIMUM (32767) */ \
	0x75, 0x10,							/*     REPORT_SIZE (16) */ \
	0x95, 0x04,							/*     REPORT_COUNT (4) */ \
	0x81, 0x02,							/*     INPUT (Data,Var,Abs) */ \
	/* 2 8bit Axis */ \
	0x09, 0x32,							/*     USAGE (Z) */ \
	0x09, 0x35,							/*     USAGE (Rz) */ \
	0x15, 0x80,							/*     LOGICAL_MINIMUM (-128) */ \
	0x25, 0x7F,							/*     LOGICAL_MAXIMUM (127) */ \
	0x75, 0x08,							/*     REPORT_SIZE (8) */ \
	0x95, 0x02,							/*     REPORT_COUNT (2) */ \
	0x81, 0x02,							/*     INPUT (Data,Var,Abs) */ \
	0xc0,								/*   END_COLLECTION */ \
	/* 2 Hat Switches */ \
	0x05, 0x01,							/*   USAGE_PAGE (Generic Desktop) */ \
	0x09, 0x39,							/*   USAGE (Hat switch) */ \
	0x09, 0x39,							/*   USAGE (Hat switch) */ \
	0x15, 0x01,							/*   LOGICAL_MINIMUM (1) */ \
	0x25, 0x08,							/*   LOGICAL_MAXIMUM (8) */ \
	0x95, 0x02,							/*   REPORT_COUNT (2) */ \
	0x75, 0x04,							/*   REPORT_SIZE (4) */ \
	0x81, 0x02,							/*   INPUT (Data,Var,Abs) */ \
	0xc0								/* END_COLLECTION */
#endif

// note by NicoHood: RawHID might never work with multireports, because of OS problems
// therefore we have to make it a single report with no idea. No other HID device will be supported then.
#define RAWHID_USAGE_PAGE	0xFFC0 // recommended: 0xFF00 to 0xFFFF
#define RAWHID_USAGE		0x0C00 // recommended: 0x0100 to 0xFFFF
#define RAWHID_TX_SIZE (USB_EP_SIZE-1)
#define RAWHID_RX_SIZE (USB_EP_SIZE-1)

#define LSB(_x) ((_x) & 0xFF)
#define MSB(_x) ((_x) >> 8)

#ifndef HID_REPORT_RAWHID
#define HID_REPORT_RAWHID(report_id) /*    RAW HID */ \
    0x06, LSB(RAWHID_USAGE_PAGE), MSB(RAWHID_USAGE_PAGE),      /* 30 */ \
    0x0A, LSB(RAWHID_USAGE), MSB(RAWHID_USAGE), \
\
    0xA1, 0x01,                  /* Collection 0x01 */ \
    0x85, report_id,			 /* REPORT_ID */ \
    0x75, 0x08,                  /* report size = 8 bits */ \
    0x15, 0x00,                  /* logical minimum = 0 */ \
    0x26, 0xFF, 0x00,            /* logical maximum = 255 */ \
\
    0x95, RAWHID_TX_SIZE,        /* report count TX */ \
    0x09, 0x01,                  /* usage */ \
    0x81, 0x02,                  /* Input (array) */ \
\
    0x95, RAWHID_RX_SIZE,        /* report count RX */ \
    0x09, 0x02,                  /* usage */ \
    0x91, 0x02,                  /* Output (array) */ \
    0xC0                         /* end collection */ 
#endif

#if defined(USBCON)

#include "USBDesc.h"
#include "USBCore.h"
// only include HIDAPI if we have an USB AVR MCU.
// The use can overwrite HID_SendReport() and manually include the APIs.
#include "HIDAPI.h"

//================================================================================
//================================================================================
//	HID 'Driver'

int		HID_GetInterface(uint8_t* interfaceNum);
int		HID_GetDescriptor(int i);
bool	HID_Setup(Setup& setup);
void	HID_SendReport(uint8_t id, const void* data, int len);
#if defined(HID_KEYBOARD_LEDS_ENABLED)
void	HID_SetKeyboardLedReport(uint8_t leds);
#endif

#else /* if defined(USBCON) */

// (weak) function prototype to let the user use the HIDAPI for different use
// if he include the api himself
void	HID_SendReport(uint8_t id, const void* data, int len);

#endif /* if defined(USBCON) */

#endif